docs(specs): Add specification for custom-builder (#14901)
influxdata/telegraf: docs/specs/tsd-002-custom-builder.md

# Telegraf Custom-Builder

## Objective

Provide a tool to build a customized, smaller version of Telegraf with only
the required plugins included.

## Keywords

tool, binary size, customization

## Overview

The Telegraf binary continues to grow as new plugins and features are added
and dependencies are updated. Users running on resource constraint systems

> such as embedded-systems or inside containers might suffer from that grow.

Reviewer: ```suggestion
such as embedded-systems or inside containers might suffer from the growth.
```

of Telegraf. Furthermore, the degree of customization and thus the effective
size reduction might vary across versions.

The specified tool will *not* produce distribution packages or containers but
*only* the customized, static Telegraf binary.

Requirements to produce a customized Telegraf binary are listed below.

### Configuration files

The user has to provide one or more valid Telegraf configuration files or
configuration directories that are used with the produced binary later on. If
you plan to use the customized binary in multiple scenarios, the configuration
files should be a superset of all expected use-cases.

### Telegraf source code

To build a customized version of Telegraf you need access to the [Telegraf
source-code repository](https://github.com/influxdata/telegraf).

### Build tools

For compiling the customized binary you need the [Golang language](https://go.dev/)
as well as the `make` build system. The minimum required version of Golang
can be found in the *Build From Source* section of the `README.md` file of your
version. Both the `go` and the `make` command must be available in your path.

> ## Workflow

Reviewer: I started going through this and this ended up more like a README than a spec. I don't think we want to document how to use the custom-builder here, but instead reduce this down to a paragraph that says, users need to clone the repo, build the custom_builder, and pass it a config.

If we do ever change anything I don't want us to have to come back over here and keep it in sync or explain why they are different.

Thoughts?

Author: Again, copied mostly from the internal document. I'm fine with moving the real examples to the README (if they are not there already) and keep this more abstract.

Reviewer: Let's do that, thanks again!

The first step is to download the Telegraf repository for the version you are
planning to customize.

```shell

> $ git clone  --branch v1.29.5 --single-branch https://github.com/influxdata/telegraf.git

Reviewer: ```suggestion
$ git clone --branch v1.29.5 --single-branch https://github.com/influxdata/telegraf.git
```

Reviewer: I'm a little concerned someone will see 1.29.5 and think that is the only version that supported this. Maybe we show building from master and mention you can check out a specific release branch?

Author: TBH I think users that take the effort to customize their builds are usually more likely to be in an enterprise environment and will most likely use some released version... As a compromise, how about explicitly showing both, maybe even with `v<version of telegraf>` instead of a concrete version number?

Cloning into 'telegraf'...
remote: Enumerating objects: 82314, done.
remote: Counting objects: 100% (82314/82314), done.
remote: Compressing objects: 100% (27977/27977), done.
remote: Total 82314 (delta 52910), reused 81858 (delta 52723), pack-reused 0
Receiving objects: 100% (82314/82314), 57.49 MiB | 11.79 MiB/s, done.
Resolving deltas: 100% (52910/52910), done.
Note: switching to '138d0d54add1e3dcd70592d069bed4218bae2bd2'.
...

$ cd telegraf
```

This will clone a specific version of Telegraf, `v1.29.5` in this case. You can

> also use the latest master or download a source tarball or zip-archive.

Reviewer: mention from the releases page? https://github.com/influxdata/telegraf/releases

# Telegraf Custom-Builder

## Objective

Provide a tool to build a customized, smaller version of Telegraf with only
the required plugins included.

## Keywords

tool, binary size, customization

## Overview

The Telegraf binary continues to grow as new plugins and features are added

> and dependencies are updated. Users running on resource constraint systems

Reviewer 2: ```suggestion
and dependencies are updated. Users running on resource constrained systems
```

Furthermore, the degree of customization and thus the effective size reduction
might vary across versions. The tool must create a single static Telegraf
binary. Distribution packages or containers are *not* targeted.

## Prior art

[PR #5809](https://github.com/influxdata/telegraf/pull/5809) and
[telegraf-lite-builder](https://github.com/influxdata/telegraf/tree/telegraf-lite-builder/cmd/telegraf-lite-builder):

- Uses docker
- Uses browser:
  - Generates a webpage to pick what options you want. User chooses plugins;
    does not take a config file
  - Build a binary, then minifies by stripping and compressing that binary
- Does some steps that belong in makefile, not builder
  - Special case for upx
  - Makes gzip, zip, tar.gz
- Uses gopkg.in?
- Can also work from the command line

[PR #8519](https://github.com/influxdata/telegraf/pull/8519)

- User chooses plugins OR provides a config file

[powers/telegraf-build](https://github.com/powersj/telegraf-build)

- User chooses plugins OR provides a config file
- Currently kept in separate repo
- Undoes changes to all.go files

[rawkode/bring-your-own-telegraf](https://github.com/rawkode/bring-your-own-telegraf)

> - Users docker

Reviewer 2: ```suggestion
- Uses docker
```
